File: app/src/main/java/io/github/neuralcoherence/probe/LiveUpdateTransition.java
package io.github.neuralcoherence.probe;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.SystemClock;

final class LiveUpdateTransition {
    private static final String PREFS_NAME = "live_update_transition";
    private static final String PREF_TOKEN = "token";
    private static final int REQUEST_CODE = 1203;

    private LiveUpdateTransition() {
    }

    static long schedule(Context sourceContext, String title, String summary,
                         String details) {
        Context context = sourceContext.getApplicationContext();
        long token = SystemClock.elapsedRealtimeNanos();
        preferences(context).edit().putLong(PREF_TOKEN, token).commit();

        Intent intent = transitionIntent(context)
                .putExtra(LiveUpdateContract.EXTRA_TRANSITION_TOKEN, token)
                .putExtra(LiveUpdateContract.EXTRA_TITLE, title)
                .putExtra(LiveUpdateContract.EXTRA_SUMMARY, summary)
                .putExtra(LiveUpdateContract.EXTRA_DETAILS, details);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(
                context, REQUEST_CODE, intent,
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
        AlarmManager manager = context.getSystemService(AlarmManager.class);
        if (manager != null) {
            manager.set(AlarmManager.ELAPSED_REALTIME_WAKEUP,
                    SystemClock.elapsedRealtime()
                            + LiveUpdateContract.COMPLETION_PREVIEW_MILLIS,
                    pendingIntent);
        }
        return token;
    }

    static void complete(Context sourceContext, String title, String summary,
                         String details, long token) {
        Context context = sourceContext.getApplicationContext();
        if (token == 0L || preferences(context).getLong(PREF_TOKEN, 0L) != token) {
            return;
        }
        cancelAlarm(context);
        preferences(context).edit().remove(PREF_TOKEN).apply();
        LiveUpdateNotification.postFinished(context, title, summary, details);
    }

    static void cancel(Context sourceContext) {
        Context context = sourceContext.getApplicationContext();
        cancelAlarm(context);
        preferences(context).edit().remove(PREF_TOKEN).apply();
    }

    private static void cancelAlarm(Context context) {
        PendingIntent pendingIntent = PendingIntent.getBroadcast(
                context, REQUEST_CODE, transitionIntent(context),
                PendingIntent.FLAG_NO_CREATE | PendingIntent.FLAG_IMMUTABLE);
        if (pendingIntent == null) {
            return;
        }
        AlarmManager manager = context.getSystemService(AlarmManager.class);
        if (manager != null) {
            manager.cancel(pendingIntent);
        }
        pendingIntent.cancel();
    }

    private static Intent transitionIntent(Context context) {
        return new Intent(context, LiveUpdateTransitionReceiver.class)
                .setAction(LiveUpdateContract.ACTION_COMPLETE_TRANSITION);
    }

    private static SharedPreferences preferences(Context context) {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }
}

File: app/src/main/kotlin/io/github/neuralcoherence/probe/LiveUpdateTransition.kt
package io.github.neuralcoherence.probe

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.SystemClock

internal object LiveUpdateTransition {
    private const val PREFS_NAME = "live_update_transition"
    private const val PREF_TOKEN = "token"
    private const val REQUEST_CODE = 1203

    fun schedule(context: Context, title: String?, summary: String?, details: String?): Long {
        val appContext = context.applicationContext
        val token = SystemClock.elapsedRealtimeNanos()
        preferences(appContext).edit().putLong(PREF_TOKEN, token).commit()
        val intent = transitionIntent(appContext)
            .putExtra(LiveUpdateContract.EXTRA_TRANSITION_TOKEN, token)
            .putExtra(LiveUpdateContract.EXTRA_TITLE, title)
            .putExtra(LiveUpdateContract.EXTRA_SUMMARY, summary)
            .putExtra(LiveUpdateContract.EXTRA_DETAILS, details)
        val pendingIntent = PendingIntent.getBroadcast(
            appContext,
            REQUEST_CODE,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
        appContext.getSystemService(AlarmManager::class.java)?.set(
            AlarmManager.ELAPSED_REALTIME_WAKEUP,
            SystemClock.elapsedRealtime() + LiveUpdateContract.COMPLETION_PREVIEW_MILLIS,
            pendingIntent,
        )
        return token
    }

    fun complete(
        context: Context,
        title: String?,
        summary: String?,
        details: String?,
        token: Long,
    ) {
        val appContext = context.applicationContext
        if (token == 0L || preferences(appContext).getLong(PREF_TOKEN, 0L) != token) return
        cancelAlarm(appContext)
        preferences(appContext).edit().remove(PREF_TOKEN).apply()
        LiveUpdateNotification.postFinished(appContext, title, summary, details)
    }

    fun cancel(context: Context) {
        val appContext = context.applicationContext
        cancelAlarm(appContext)
        preferences(appContext).edit().remove(PREF_TOKEN).apply()
    }

    private fun cancelAlarm(context: Context) {
        val pendingIntent = PendingIntent.getBroadcast(
            context,
            REQUEST_CODE,
            transitionIntent(context),
            PendingIntent.FLAG_NO_CREATE or PendingIntent.FLAG_IMMUTABLE,
        ) ?: return
        context.getSystemService(AlarmManager::class.java)?.cancel(pendingIntent)
        pendingIntent.cancel()
    }

    private fun transitionIntent(context: Context): Intent =
        Intent(context, LiveUpdateTransitionReceiver::class.java)
            .setAction(LiveUpdateContract.ACTION_COMPLETE_TRANSITION)

    private fun preferences(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}
